core/src/main/java/org/bitcoinj/core/Transaction.java

```diff
@@ -1413,67 +1412,64 @@ public synchronized Sha256Hash hashForWitnessSignature(
+        Sha256Hash hashOutputs = Sha256Hash.ZERO_HASH;
+        int basicSigHashType = sigHashType & 0x1f;
+        boolean anyoneCanPay = (sigHashType & SigHash.ANYONECANPAY.value) == SigHash.ANYONECANPAY.value;
+        boolean signAll = (basicSigHashType != SigHash.SINGLE.value) && (basicSigHashType != SigHash.NONE.value);
+
+        if (!anyoneCanPay) {
+            ByteBuffer bufHashPrevouts = ByteBuffer.allocate(this.inputs.size() * (Sha256Hash.LENGTH + 4));
+            for (TransactionInput input : this.inputs) {
+                input.getOutpoint().hash().write(bufHashPrevouts);
+                writeInt32LE(input.getOutpoint().index(), bufHashPrevouts);
             }
+            hashPrevouts = Sha256Hash.wrapReversed(Sha256Hash.hashTwice(bufHashPrevouts.array()));
+        }
 
-            if (!anyoneCanPay && signAll) {
-                ByteArrayOutputStream bosSequence = new ByteArrayOutputStream(256);
-                for (TransactionInput input : this.inputs) {
-                    writeInt32LE(input.getSequenceNumber(), bosSequence);
-                }
-                hashSequence = Sha256Hash.hashTwice(bosSequence.toByteArray());
+        if (!anyoneCanPay && signAll) {
+            ByteBuffer bufSequence = ByteBuffer.allocate(this.inputs.size() * 4);
+            for (TransactionInput input : this.inputs) {
+                writeInt32LE(input.getSequenceNumber(), bufSequence);
             }
+            hashSequence = Sha256Hash.wrapReversed(Sha256Hash.hashTwice(bufSequence.array()));
+        }
 
-            if (signAll) {
-                ByteArrayOutputStream bosHashOutputs = new ByteArrayOutputStream(256);
-                for (TransactionOutput output : this.outputs) {
-                    writeInt64LE(output.getValue().getValue(), bosHashOutputs);
-                    byte[] scriptBytes = output.getScriptBytes();
-                    bosHashOutputs.write(VarInt.of(scriptBytes.length).serialize());
-                    bosHashOutputs.write(scriptBytes);
-                }
-                hashOutputs = Sha256Hash.hashTwice(bosHashOutputs.toByteArray());
-            } else if (basicSigHashType == SigHash.SINGLE.value && inputIndex < outputs.size()) {
-                ByteArrayOutputStream bosHashOutputs = new ByteArrayOutputStream(256);
-                writeInt64LE(this.outputs.get(inputIndex).getValue().getValue(), bosHashOutputs);
-                byte[] scriptBytes = this.outputs.get(inputIndex).getScriptBytes();
-                bosHashOutputs.write(VarInt.of(scriptBytes.length).serialize());
-                bosHashOutputs.write(scriptBytes);
-                hashOutputs = Sha256Hash.hashTwice(bosHashOutputs.toByteArray());
+        if (signAll) {
+            ByteBuffer bufHashOutputs = ByteBuffer.allocate(this.outputs.stream().mapToInt(
+                    output -> Coin.BYTES + Buffers.lengthPrefixedBytesSize(output.getScriptBytes())
+            ).sum());
```

Comment:
This length calculation should be moved to a private method and could be done in a separate commit (prior to the BAOS -> ByteBuffer conversion)